### ISSUE TITLE
BufferingClientHttpResponseWrapper overridden in MockRestServiceServer

### ISSUE THREAD

**Issue by sddakoty** [spring-projects/spring-boot#12494]
When using a `RestTemplate` with `BufferingClientHttpRequestFactory`, this will create a `BufferingClientHttpResponseWrapper` that allows reading the webservice response multiple times.

Useful for logging.

Problem: during a junit `@Test` when using a `MockRestServiceServer`, the factory seams to be replaced and thus, if the logging interceptor is still configured, the response returned by the mocked `RestTemplate` is always `null`.

There are several (pending or closed) issues or workarounds for that problem that nobody seems to be working on:

https://github.com/spring-projects/spring-boot/issues/6686
https://dev.to/shriharshmishra/today-i-learned-to-use-custom-clienthttprequestfactory-with-mockrestserviceserver--3bo
https://jira.spring.io/browse/SPR-14694

Maybe somebody can have a deeper look into it and fix this in `spring-test` or `spring-boot`?

```
public class RestInterceptor implements ClientHttpRequestInterceptor {
	@Override
	public ClientHttpResponse intercept(HttpRequest request, byte[] body,
	                                    ClientHttpRequestExecution execution) throws IOException {
		ClientHttpResponse response = execution.execute(request, body);
		LOGGER.info(IOUtils.toString(rsp.getBody())); //this empties the body in mocked setup
		return response;
	}
}


@Configuration
public class AppConfig {
	//by using BufferingClientHttpRequestFactory, we initialize BufferingClientHttpResponseWrapper
	@Bean
	public RestTemplate restTemplate(RestTemplateBuilder builder) {
		HttpComponentsClientHttpRequestFactory factory = new HttpComponentsClientHttpRequestFactory(HttpClientBuilder.create().build());

		return builder.requestFactory(new BufferingClientHttpRequestFactory(factory))
				.additionalInterceptors(Arrays.asList(new RestInterceptor())
				.build();
	}
}
```

In a normal spring environment, the BufferingClientHttpResponseWrapper will enable reading the response body multiple times.

There is actually a **workaround** (overriding the `BufferingClientHttpRequestFactory` of `RestTemplate` in mock setup), but it's probably better this could be fixed in future:

```
	private MockRestServiceServer mockServer;

	@Autowired
	private RestTemplate restTemplate;

	@Before
	public void createWebserver() throws Exception {
		mockServer = MockRestServiceServer.bindTo(restTemplate).build();

                //WORKAROUND
		final ClientHttpRequestFactory requestFactory =
				(ClientHttpRequestFactory) ReflectionTestUtils.getField(restTemplate, "requestFactory");
		restTemplate.setRequestFactory(new BufferingClientHttpRequestFactory(requestFactory));
	}
```

**Comment by wilkinsona** [comment]
Please don't deliberately raise duplicate issues. As you've noted, there's already a Spring Framework [issue](https://jira.spring.io/browse/SPR-14694) tracking this problem. @rstoyanchev commented on that issue looking for some clarification but didn't get any. If you're in a position to recreate the problem and give Rossen the information that he needs to understand the problem and hopefully be able to fix it, please do so by commenting on the JIRA ticket.

**Comment by fstorz** [comment]
solved by [#19258](https://github.com/spring-projects/spring-framework/issues/19258)
